### PR TITLE
Added Vlan discovery on LAG ports, Jetstream OS

### PR DESCRIPTION
**Pull request by Npeca75**
Vlans discovery on LAG ports on TP-LINK Jetstram OS was missing

DO NOT DELETE THE UNDERLYING TEXT

#### Please note

> Please read this information carefully. You can run `./lnms dev:check` to check your code before submitting.

- [x] Have you followed our [code guidelines?](https://docs.librenms.org/Developing/Code-Guidelines/)
- [ ] If my Pull Request does some changes/fixes/enhancements in the WebUI, I have inserted a screenshot of it.

#### Testers

If you would like to test this pull request then please run: `./scripts/github-apply <pr_id>`, i.e `./scripts/github-apply 5926`
After you are done testing, you can remove the changes with `./scripts/github-remove`.  If there are schema changes, you can ask on discord how to revert.

### CODE REVIEW

**Comment by PipoCanaja** [comment]
Hi @Npeca75 
Please check the style tests and update your code accordingly : 
https://github.styleci.io/analyses/Vrpa53?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

And please provide test files (snmprec and JSON) as required. 

Thanx

**Comment by Npeca75** [comment]
Hi @PipoCanaja 

sorry for mess, i am totally noob on GITHUB
looks like code check is OK
attached the snmpwalk from Jetsream switch with LAG's
and a screenshot, clearly visible that Vlans are added on Po1, Po2 ...

Screenshot_20210708_164311

[Jetstream-LAG.walk.txt](https://github.com/librenms/librenms/files/6785409/Jetstream-LAG.walk.txt)

File: includes/discovery/vlans/jetstream.inc.php [hunk 1]
```diff
@@ -29,14 +29,34 @@
...
+// 2021-06-07: Added Vlan parsing on LAG ports
 //
 
 if (! function_exists('jetstreamExpand')) {
     function jetstreamExpand($var)
     {
+        unset($result);
+
+        if ($pos = strpos($var, ',LAG')) {         // if input string contains 'LAG'
+            $lag = str_replace('LAG', '', substr($var, $pos + 1)); // get the LAG ports in form of '1,2,4-8'
```
Comment (murrant) [inline includes/discovery/vlans/jetstream.inc.php:41]: This code is really hard to read, could it perhaps be done with a regex?

Comment (Npeca75) [reply includes/discovery/vlans/jetstream.inc.php:41]: hi @murrant 

sorry, my "native" language is assembler on embedded devices, so i am not familiar with PHP
if you have any suggestion ...
TP-LINK made own way of representing Vlans in snmp
for example:
STRING: "1/0/1-2,1/0/4-6,1/0/25,LAG1-2,LAG4"
as you see, port 1,2,4,5,6,25 are tagged
and LAG1,2,4 are tagged
so, the information about ports and LAG's are in one line
it need to be splitted on two parts, first for ports, second for LAG's
then DASH need to be replaced with corresponding numbers
4-6 -> 4,5,6
same technique for PORTS (left) part of string
and same for LAG's (right) part of string

anyway, if you could / want to rewrite this i will be happy

Comment (murrant) [reply includes/discovery/vlans/jetstream.inc.php:41]: Try using a site like regex101.com  https://regex101.com/r/7utMDz/1

Sample code:
```php
$string = '1/0/1-2,1/0/4-6,1/0/25,LAG1-2,LAG4';
preg_match_all('/LAG(\d+)(?:-(\d+))?/', $string, $matches);
foreach ($matches[1] as $index => $start) {
    $end = $matches[2][$index] ?: $start;
    for ($i = $start; $i <= $end; $i++) {
        dump((int)$i);
    }
}
```

**Comment by Npeca75** [comment]
> Hi @Npeca75
> 
> And please provide test files (snmprec and JSON) as required.
> 
Ok, i hope these files are enough for testing

[jetstream.snmprec.txt](https://github.com/librenms/librenms/files/6793997/jetstream.snmprec.txt)
[jetstream.snmpwalk.txt](https://github.com/librenms/librenms/files/6793998/jetstream.snmpwalk.txt)

File: includes/discovery/vlans/jetstream.inc.php (hunk 1 above)
Comment (Npeca75) [reply includes/discovery/vlans/jetstream.inc.php]: hi @murrant 
following your advice , the whole function was rewritten ...

    function jetstreamExpand($var)
    {
        unset($result);

        preg_match_all('/LAG(\d+)(?:-(\d+))?/', $var, $lags);

        foreach ($lags[1] as $index => $start) {
            $end = $lags[2][$index] ?: $start;
            for ($i = $start; $i <= $end; $i++) {
                $result[] = 'LAG' . $i; //need to be in form LAGx
            }
        }

        preg_match_all('/1\/0\/(\d+)(?:-(\d+))?/', $var, $ports);
        foreach ($ports[1] as $index => $start) {
            $end = $ports[2][$index] ?: $start;
            for ($i = $start; $i <= $end; $i++) {
                $result[] = '1/0/' . $i; //need to be in form 1/0/x
            }
        }

        return $result;
    }

**Comment by Jellyfrog** [comment]
> Please add test data so we can ensure your change is not broken in the future.
> 
> Read the docs to find out how: https://docs.librenms.org/Developing/os/Test-Units

Please see this on how to generate it. And it needs to be commited also

**Comment by Npeca75** [comment]
hi @Jellyfrog 
 > Please add test data so we can ensure your change is not broken in the future.
ok
./scripts/collect-snmp-data.php -h 1 -v t1600g-28ts

and pushed to GIT

**Comment by murrant** [comment]
Looks much better @Npeca75 I've generated the test data for you and uploaded it. Here is the command (since vlans is not enabled by default)
`./scripts/save-test-data.php -o jetstream -v t1600g-28ts -m vlans`

File: includes/discovery/vlans/jetstream.inc.php [hunk 2]
```diff
@@ -29,26 +29,29 @@
...
-                $tmp = explode('-', $element); // left part is a fully defined port, right is the end number of the serie
-                $port_start_array = explode('/', $tmp[0]);
-                $port_id = trim(array_pop($port_start_array)); // $port_start_array is "[x, y]", $port_id is "a";
 
-                for ($i = $port_id; $i <= $tmp[1]; $i++) {
-                    $result[] = implode('/', array_merge($port_start_array, [$i]));
-                }
-            } else {
-                $result[] = $element;
+        preg_match_all('/LAG(\d+)(?:-(\d+))?/', $var, $lags);
```
Comment (murrant) [inline includes/discovery/vlans/jetstream.inc.php:41]: Hmm, we could combine these two regex. :)

**Comment by Npeca75** [comment]
hi @murrant 

i get this error running snmpsim

> ./scripts/save-test-data.php -o jetstream -v t1600g-28ts -m vlans

Starting snmpsim listening on 127.1.6.1:1161...
OS: jetstream
Module: vlans
Variant: t1600g-28ts

jetstream_t1600g-28ts: Could not connect to 127.1.6.1, please check the snmp details and snmp reachability

**Review by murrant — APPROVED** [review]
Nice :)

**Comment by murrant** [comment]
@Npeca75 it requires you have a working snmpsim on your system.

**Pull request merged**